### PR TITLE
feat: autostop workspaces owned by suspended users 

### LINKED ISSUE #8051: Stop a user's workspaces when their account is suspended
When a user is suspended, their API key is invalidated and they cannot establish any new connections. However, if a user happens to have an active SSH connection, they can remain in their workspace to download files, access network resources, etc.

It seems like are two ways this could work, both are fine with me:
- Disconnect all open connections from the user (including direct peer-to-peer connections)
- Stop the workspace, which will inherently disconnect all connections.

### PR DESCRIPTION
**Pull request by mtojek**
Fixed: https://github.com/coder/coder/issues/8051

This PR modifies autostop rules to stop workspaces owned by suspended users.

### CODE REVIEW

File: coderd/autobuild/lifecycle_executor_test.go [hunk 1]
```diff
@@ -563,6 +563,53 @@ func TestExecutorWorkspaceAutostopBeforeDeadline(t *testing.T) {
...
+	_, err := client.UpdateUserStatus(ctx, user.ID.String(), codersdk.UserStatusSuspended)
+	require.NoError(t, err, "update user status")
+
+	// When: the autobuild executor ticks after the scheduled time
+	go func() {
+		tickCh <- sched.Next(workspace.LatestBuild.CreatedAt)
+		close(tickCh)
+	}()
+
+	// Then: nothing should happen
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor_test.go:602]: bad copy-paste, fixed

Comment (dannykopping) [inline coderd/autobuild/lifecycle_executor_test.go:602]: This is a bit misleading. Can you rephrase?

File: coderd/autobuild/lifecycle_executor_test.go [hunk 2]
```diff
@@ -563,6 +563,53 @@ func TestExecutorWorkspaceAutostopBeforeDeadline(t *testing.T) {
...
+	// Given: workspace is started, and the user is suspended.
+	workspace = coderdtest.MustWorkspace(t, userClient, workspace.ID)
+	_, err := client.UpdateUserStatus(ctx, user.ID.String(), codersdk.UserStatusSuspended)
+	require.NoError(t, err, "update user status")
+
+	// When: the autobuild executor ticks after the scheduled time
+	go func() {
+		tickCh <- sched.Next(workspace.LatestBuild.CreatedAt)
+		close(tickCh)
+	}()
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor_test.go:600]: I assume that the lifecycle executor stops the workspace right after the conditional logic is satisfied? Do you have a specific idea in mind on how we can improve testing?

File: coderd/autobuild/lifecycle_executor_test.go [hunk 3]
```diff
@@ -563,6 +563,53 @@ func TestExecutorWorkspaceAutostopBeforeDeadline(t *testing.T) {
...
+
+	// Then: nothing should happen
+	stats := <-statsCh
+	assert.Len(t, stats.Errors, 0)
+	assert.Len(t, stats.Transitions, 1)
+	assert.Equal(t, stats.Transitions[workspace.ID], database.WorkspaceTransitionStop)
+
+	// Wait for stop to complete
+	workspace = coderdtest.MustWorkspace(t, client, workspace.ID)
+	_ = coderdtest.AwaitWorkspaceBuildJobCompleted(t, client, workspace.LatestBuild.ID)
```
Comment (dannykopping) [inline coderd/autobuild/lifecycle_executor_test.go:610]: We should probably validate here that the build transition is the one we expect.

File: coderd/autobuild/lifecycle_executor.go [hunk 4]
```diff
@@ -387,6 +387,10 @@ func isEligibleForAutostop(ws database.Workspace, build database.WorkspaceBuild,
 		return false
 	}
 
+	if build.Transition == database.WorkspaceTransitionStart && user.Status == database.UserStatusSuspended {
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor.go:390]: No, I don't want to mess with others.

File: coderd/autobuild/lifecycle_executor_test.go [hunk 5]
```diff
@@ -563,6 +563,53 @@ func TestExecutorWorkspaceAutostopBeforeDeadline(t *testing.T) {
...
+	coderdtest.AwaitTemplateVersionJobCompleted(t, client, version.ID)
+	template := coderdtest.CreateTemplate(t, client, admin.OrganizationID, version.ID)
+	userClient, user := coderdtest.CreateAnotherUser(t, client, admin.OrganizationID)
+	workspace := coderdtest.CreateWorkspace(t, userClient, admin.OrganizationID, template.ID, func(cwr *codersdk.CreateWorkspaceRequest) {
+		cwr.AutostartSchedule = ptr.Ref(sched.String())
+	})
+	coderdtest.AwaitWorkspaceBuildJobCompleted(t, userClient, workspace.LatestBuild.ID)
+
+	// Given: workspace is started, and the user is suspended.
+	workspace = coderdtest.MustWorkspace(t, userClient, workspace.ID)
```
Comment (dannykopping) [inline coderd/autobuild/lifecycle_executor_test.go:592]: We should validate here that the workspace is in the expected state before we do the rest.

File: coderd/autobuild/lifecycle_executor.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/autobuild/lifecycle_executor.go:390]: I was going to mention `dormant` but figured it was out of scope; _should_ dormant users be allowed to run workloads?

File: coderd/autobuild/lifecycle_executor_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (johnstcn) [reply coderd/autobuild/lifecycle_executor_test.go:600]: No matter what time you send down the channel, it should result in the workspace being stopped.

File: coderd/autobuild/lifecycle_executor.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor.go:390]: IIRC on the first interaction with API the dormant user will be switched to active, so yes.

[the remark this answers is not in the document]
Comment (johnstcn) [reply coderd/autobuild/lifecycle_executor.go:390]: By definition... they shouldn't be able to...?

File: coderd/autobuild/lifecycle_executor_test.go (hunk 5 above)
Comment (johnstcn) [reply coderd/autobuild/lifecycle_executor_test.go:592]: `AwaitWorkspaceBuildJobCompleted` should do that for us.

Comment (dannykopping) [reply coderd/autobuild/lifecycle_executor_test.go:592]: AFAICS that just shows that the build completed, but not that it's in the desired state.

Comment (johnstcn) [reply coderd/autobuild/lifecycle_executor_test.go:592]: Good call actually :+1:

File: coderd/autobuild/lifecycle_executor.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor.go:390]: Ok, what is your preference - should I extend the logic to dormant users too?

[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/autobuild/lifecycle_executor.go:390]: I think it could be a follow-up PR since there may be additional considerations.

File: coderd/autobuild/lifecycle_executor_test.go (hunk 3 above)
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor_test.go]: Done 👍

File: coderd/autobuild/lifecycle_executor_test.go (hunk 5 above)
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor_test.go:592]: Done 👍

**Review by dannykopping — APPROVED** [review]
LGTM

File: coderd/autobuild/lifecycle_executor_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor_test.go]: Thanks for more context. I excluded `sched` from the test.

**Review by johnstcn — APPROVED** [review]
